### DISCUSSION TITLE
Getting Started -   I can't install gfortran using brew.

### DISCUSSION THREAD

**Discussion by FelixTheKitten** [Homebrew/discussions#4695] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: 128df14e0afd0525564051a7665708c2973a3cab
Last commit: 4 days ago
Core tap JSON: 04 Aug 06:36 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 12.6.6-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

If I try "brew install gcc" or "brew install gfortran", the installation fails.  

It spits out a whole bunch of stuff along the lines of:

cp: /private/tmp/d20230803-2486-1h3xx5o/xz/./5.4.3/share/doc: unable to copy ACL to /usr/local/Cellar/xz/./5.4.3/share/doc: No such file or directory
cp: utimensat: /usr/local/Cellar/xz/./5.4.3/share: No such file or directory
cp: chown: /usr/local/Cellar/xz/./5.4.3/share: No such file or directory
cp: chmod: /usr/local/Cellar/xz/./5.4.3/share: No such file or directory
cp: chflags: /usr/local/Cellar/xz/./5.4.3/share: No such file or directory
cp: /private/tmp/d20230803-2486-1h3xx5o/xz/./5.4.3/share: unable to copy ACL to /usr/local/Cellar/xz/./5.4.3/share: No such file or directory
cp: utimensat: /usr/local/Cellar/xz/./5.4.3: No such file or directory
cp: chown: /usr/local/Cellar/xz/./5.4.3: No such file or directory
cp: chmod: /usr/local/Cellar/xz/./5.4.3: No such file or directory
cp: chflags: /usr/local/Cellar/xz/./5.4.3: No such file or directory
cp: /private/tmp/d20230803-2486-1h3xx5o/xz/./5.4.3: unable to copy ACL to /usr/local/Cellar/xz/./5.4.3: Permission denied
cp: utimensat: /usr/local/Cellar/xz/.: Permission denied

**Comment by osalbahr** [comment]
Not sure. Could you provide the full output?

Also, try `brew install --verbose --debug <formula>` that might give some clues.

**Reply by FelixTheKitten** [reply]
Here is the first part of the output showing where things go wrong.  Everything seems fine until it tries  Installing gcc dependency: xz
==> Pouring xz--5.4.3.monterey.bottle.tar.gz.

brew install gfortran
==> Downloading https://formulae.brew.sh/api/formula.jws.json
######################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
######################################################################### 100.0%
==> Fetching dependencies for gcc: gmp, isl, mpfr, libmpc, lz4, xz and zstd
==> Fetching gmp
==> Downloading https://ghcr.io/v2/homebrew/core/gmp/manifests/6.2.1_1
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gmp/blobs/sha256:dddc6d8c871c92
######################################################################### 100.0%
==> Fetching isl
==> Downloading https://ghcr.io/v2/homebrew/core/isl/manifests/0.26
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/isl/blobs/sha256:0dcc555fd2517c
######################################################################### 100.0%
==> Fetching mpfr
==> Downloading https://ghcr.io/v2/homebrew/core/mpfr/manifests/4.2.0-p12
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/mpfr/blobs/sha256:d6879fb531aac
######################################################################### 100.0%
==> Fetching libmpc
==> Downloading https://ghcr.io/v2/homebrew/core/libmpc/manifests/1.3.1
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libmpc/blobs/sha256:c32f2c3fe7a
######################################################################### 100.0%
==> Fetching lz4
==> Downloading https://ghcr.io/v2/homebrew/core/lz4/manifests/1.9.4
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/lz4/blobs/sha256:88b369cea90a0a
######################################################################### 100.0%
==> Fetching xz
==> Downloading https://ghcr.io/v2/homebrew/core/xz/manifests/5.4.3
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/xz/blobs/sha256:1197fc939b8c972
######################################################################### 100.0%
==> Fetching zstd
==> Downloading https://ghcr.io/v2/homebrew/core/zstd/manifests/1.5.5-1
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/zstd/blobs/sha256:0d57c7e4a4735
######################################################################### 100.0%
==> Fetching gcc
==> Downloading https://ghcr.io/v2/homebrew/core/gcc/manifests/13.1.0
######################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gcc/blobs/sha256:f9cbc7eb14781d
######################################################################### 100.0%
==> Installing dependencies for gcc: gmp, isl, mpfr, libmpc, lz4, xz and zstd
==> Installing gcc dependency: gmp
==> Pouring gmp--6.2.1_1.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/gmp/6.2.1_1: 21 files, 3.3MB
==> Installing gcc dependency: isl
==> Pouring isl--0.26.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/isl/0.26: 73 files, 7.6MB
==> Installing gcc dependency: mpfr
==> Pouring mpfr--4.2.0-p12.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/mpfr/4.2.0-p12: 30 files, 3MB
==> Installing gcc dependency: libmpc
==> Pouring libmpc--1.3.1.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/libmpc/1.3.1: 12 files, 464KB
==> Installing gcc dependency: lz4
==> Pouring lz4--1.9.4.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/lz4/1.9.4: 22 files, 685.1KB
==> Installing gcc dependency: xz
==> Pouring xz--5.4.3.monterey.bottle.tar.gz
cp: /usr/local/Cellar/xz/./5.4.3: Permission denied
cp: /private/tmp/d20230803-2486-1h3xx5o/xz/./5.4.3: unable to copy extended attributes to /usr/local/Cellar/xz/./5.4.3: Permission denied

**Comment by SMillerDev** [comment]
What are the permissions for `/usr/local/Cellar/xz/`?

**Reply by FelixTheKitten** [reply]
There are two accounts on this computer.  The first one I created was stephensalser.  Then I created an alternate account called ssalserold to hold files from a failing laptop.

The xz directory seems to be owned by ssalserold, but I am logged in to stephensalser.

stephensalser@Stephens-MacBook-Pro-2 xz % cd /usr/local/Cellar/   
stephensalser@Stephens-MacBook-Pro-2 Cellar % ls -l
total 0
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 gdbm
drwxr-xr-x  3 stephensalser  admin  96 Aug  4 10:50 gmp
drwxr-xr-x  3 stephensalser  admin  96 Aug  4 10:50 isl
drwxr-xr-x  3 stephensalser  admin  96 Aug  4 10:50 libmpc
drwxr-xr-x  3 stephensalser  admin  96 Aug  4 10:50 lz4
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 mpdecimal
drwxr-xr-x  3 stephensalser  admin  96 Aug  4 10:50 mpfr
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 openssl@1.1
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 pkg-config
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 python@3.9
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 readline
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 sqlite
drwxr-xr-x  3 ssalserold     admin  96 Jun 24  2021 xz

**Reply by osalbahr** [reply]
Are you trying to use Homebrew multi-user or want to be able to use it only in the current user? 

If it’s the latter, try to `sudo chown -R $(whoami) /usr/local/Cellar`. (Or, uninstall and reinstall Homebrew).

(edit: spelling lol)

You might have to recursively `chown` multiple/all Homebrew directories. See https://github.com/orgs/Homebrew/discussions/3889#discussioncomment-4892469.

If it’s the former, I don’t know if Homebrew natively supports that on macOS. But I know there are workarounds.

**Reply by FelixTheKitten** [reply]
Thanks for the advice.   The command you gave fixed most of the installation process: (   sudo chown -R $(whoami) /usr/local/Cellar.   I had to fix the spelling of Cellar, but it worked! )
 
But when it gets to the very end, it still has a problem with symlinking.  Here is the error message:

Pouring python@3.9--3.9.17_1.monterey.bottle.tar.gz
Error: An unexpected error occurred during the `brew link` step
The formula built, but is not symlinked into /usr/local
Permission denied @ apply2files - /usr/local/Frameworks/Python.framework/Versions/3.9
Error: Permission denied @ apply2files - /usr/local/Frameworks/Python.framework/Versions/3.9

Do I need to check the file permissions in /usr/local/Frameworks/Python.framework/Versions/3.9 ?

**Reply by osalbahr** [reply]
> Do I need to check the file permissions in /usr/local/Frameworks/Python.framework/Versions/3.9 ?

Yeah, probably. You should have write access to all directories under `/usr/local`. You can read more about the implementation [in this stackexchange](https://security.stackexchange.com/a/191614) (linked in https://en.m.wikipedia.org/wiki/Homebrew_(package_manager)#Implementation).

Who owns `/usr/local/Frameworks`?

- - -

It might be easier to just re-run the install script. It sets permission as needed (hopefully) and I don’t think it deletes anything. You might want to backup `/usr/local` just in case tho.

**Comment by FelixTheKitten** [comment]
The owner is now "502"


I deleted the 2nd user account on the computer thinking that would get rid of the offending files.
Unfortunately, the files are still there, but the user name has been changed to 502.  Apparently this is Apple's shorthand for 2nd admin user account -- 501 would be shorthand for the primary account.


I am puzzling over whether it would be safe to simply delete all the files owned by 502, or whether this might destroy the operating system.

I am not even sure I know how to delete them all.   

I can find them all using   
sudo find / -user 502

But I don't know what -exec line I would put after this to delete each of the files.

Thanks for your help, by the way!






    On Friday, August 4, 2023 at 02:27:36 PM PDT, Osama Albahrani ***@***.***> wrote:  
 
 


Who owns /usr/local/Frameworks?

**Reply by osalbahr** [reply]
> I am puzzling over whether it would be safe to simply delete all the files owned by 502, or whether this might destroy the operating system.

You can. I don’t think the system puts system files under `/usr/local` and iirc it’s initially empty, so it’s probably fine system-wise. If it makes you feel better, macOS is an immutable operating system so it’s difficult to break the system unless you try really hard.

That being said, I think it’s better to just `chown` the files instead to avoid breakage of userspace programs. To me, it’s easier to `chown` the directory recursively than hack with `find` (I’ve never used `-exec`, I prefer to `| xargs` the output or use `$(find …)` process substitution. After making sure there are no spaces).

- - -

 If you just want to reset the state of Homebrew, the official [uninstall script](https://github.com/homebrew/install#uninstall-homebrew) is your friend.

**Comment by FelixTheKitten** [comment]
Hmmm...
deleting these files or chowning these files are both irreversible options.

Would it be possible to modify the path on all these files to effectively group them all?
For example, create a path /mess.   and modify the path to all these folders by appending /mess to the beginning of all the file paths?



    On Friday, August 4, 2023 at 03:56:24 PM PDT, Osama Albahrani ***@***.***> wrote:  
 
 



I am puzzling over whether it would be safe to simply delete all the files owned by 502, or whether this might destroy the operating system.


You can. I don’t think the system puts system files under /usr/local and iirc it’s initially empty, so it’s probably fine system-wise. If it makes you feel better, macOS is an immutable operating system so it’s difficult to break the system unless you try really hard.

That being said, I think it’s better to just chown the files instead to avoid breakage of userspace programs.

If you just want to reset the state of Homebrew, the official uninstall script is your friend.

**Reply by osalbahr** [reply]
> deleting these files or chowning these files are both irreversible options.

Yes.

> Would it be possible to modify the path on all these files to effectively group them all?
For example, create a path /mess.   and modify the path to all these folders by appending /mess to the beginning of all the file paths?

If you mean modify the symlinks, yes you can, but where would you point the symlinks to? They should be pointed to where Homebrew installs stuff. The problem, it seems, is that the directories where Homebrew installs stuff are not fully owned by you. So, running a `brew` command expectedly gets permission errors. Afterall, `brew` runs as you, not as `root`. The only way to fix this is to make sure you (and consequently `brew`) have write access to all that’s under `/usr/local/*`.
